source/geod24/concurrency.d

```diff
@@ -1503,6 +1503,152 @@ private:
     assert(x[0] == 5);
 }
 
+/***********************************************************************
+
+    A common interface for objects that can be used in `select()`
+
+***********************************************************************/
+
+public interface Selectable
+{
+    void selectWrite (void* ptr, SelectState sel_state, int sel_id);
+    void selectRead (void* ptr, SelectState sel_state, int sel_id);
+}
+
+/***********************************************************************
+
+    An aggregate to hold neccessary information for a select operation
+
+***********************************************************************/
+
+public struct SelectEntry
+{
+    /// Reference to a Selectable object
+    Selectable selectable;
+
+    /// Pointer to the select data
+    void* select_data;
+
+    this (Selectable selectable, void* select_data)
+    {
+        this.selectable = selectable;
+        this.select_data = select_data;
+    }
+}
+
+public alias SelectReturn = Tuple!(bool, "success", int, "id");
+
+/***********************************************************************
+
+    Block on multiple `Channel`s
+
+    Only one operation is completed per `select()` call
+
+    Params:
+        read_list = List of `Channel`s to select for read operation
+
+        write_list = List of `Channel`s to select for write operation
+
+    Return:
+        Returns success/failure status of the operation and the index
+        of the `Channel` that the operation was carried on
+
+***********************************************************************/
+
+public SelectReturn select (ref SelectEntry[] read_list, ref SelectEntry[] write_list)
+{
+    auto ss = new SelectState();
+    int sel_id = 0;
+
+    foreach(ref entry; read_list)
+    {
+        entry.selectable.selectRead(entry.select_data, ss, sel_id++);
+    }
+
```

Comment:
Does `sel_id` maybe have to be reset to 0 here before the second foreach loop?

Because if so, I would suggest to just use `foreach (sel_id, ref entry; *_list)` in both foreach loops.

Reply:
No, `sel_id` returned is the index of the SelectEntry in the `read_list ~ write_list`. It should be documented better.